File: 2.2.H/UNO/src/UNO.java
import java.util.EmptyStackException;
import java.util.List;

public class UNO {

    private Deck deck;
    private Dask dask = new Dask();
    private List<Player> players;
    private Player winner;

    public UNO() {
        setDeck();
    }

    private void setDeck() {
        this.deck = Deck.standard40Cards();
        deck.setGame(this);
    }

    public void joinPlayer(List<Player> players) {
        this.players = players;
        players.forEach(player -> player.setGame(this));
    }

    public void init() {
        for (int p = 0 ; p<players.size() ; p++){
            players.get(p).nameSelf(p+1);
        }
        // players.forEach(player -> player.nameSelf());
        deck.shuffle();
        for(Player player : players) {
            for(int i = 0; i < 5; i++) {
                playerAddCard(player);
            }
        }
        Dask.addCard(deck.drawCard());
    }

    public void playerAddCard(Player player) throws EmptyStackException {
        if (deck.getCardAmount() == 0){
            throw new EmptyStackException();
        }
        player.addCard(deck.drawCard());
    }

    public void takeTurn() {
        while (winner == null)
            for(Player player : players) {
                if(player.getHands().size() == 0) {
                    System.out.println(player.getName() + " wins!");
                    setWinner(player);
                    return;
                }
                if(deck.getCardAmount() == 0) {
                    deck.addCard(dask.drawCardToDeck());
                }
                Card showedCard = player.showCard(dask.showTopCard());
                if(showedCard != null)
                    dask.addCard(showedCard);
            }
    }

    private void setWinner(Player player) {
        this.winner = player;
    }

}

File: 2.2.H/python/uno.py
from deck import Deck
from player import Player
from traitlets import Instance


class UNO:
    def __init__(self):
        self.__players = list()
        self.__deck = Deck()
        self.__desk = list()
        self.__winner = Instance(Player)

    def join_player(self, player):
        if not isinstance(player, Player):
            raise ValueError("Only players can join the game")

        if len(self.__players) > 4:
            raise ValueError("The game is full")

        self.__players.append(player)
        player.game = self

    def start(self):

        for player in self.__players:
            player.name_self()

        self.__deck.shuffle()

        for player in self.__players:
            for _ in range(5):
                drawed_card = self.__deck.draw_card()
                player.add_card(drawed_card)

        self.__desk.append(self.__deck.draw_card())

    def take_turn(self):
        for player in self.__players:
            if player.has_no_hand():
                self.__winner = player
                return

            top_card = self.__desk[-1]
            player.show_card(top_card)

            if self.__deck.is_empty():
                self.refill_to_deck()

    def refill_to_deck(self):
        top_card = self.__desk.pop()
        self.__deck.refill(self.__desk)
        self.__desk = [top_card]

